### PR TITLE
Schema Validation Refactors

### PR DESCRIPTION
**Pull request by lurraca**
## Why
The SAPI Team was not happy with the current implementation and abstraction of the Catalog Plan Schema Validation piece of work in Cloud Controller.

## What
Refactor and improve the abstraction around Catalog Plan Schema Validation.

Changed the order that Schema Validations run.

## Notes
This PR doesn't change any important functionality in Cloud Controller, except for the order that Schema Validations are run.

## PR
* [X] I have viewed signed and have submitted the Contributor License Agreement
* [X] I have made this pull request to the `master` branch
* [X] I have run all the unit tests using `bundle exec rake`
* [X] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats) on bosh lite
Luis & @Ablease

### CODE REVIEW

File: lib/services/service_brokers/service_manager.rb [hunk 1]
```diff
@@ -58,16 +58,34 @@ def update_or_create_plans(catalog)
           plan.public = false
         end
 
+        create_instance = nil
+        begin
+          create_instance = catalog_plan.schemas.service_instance.create.schema.try(:to_json)
```
Comment (ericpromislow) [inline lib/services/service_brokers/service_manager.rb]: `create_instance = catalog_plan.schemas.service_instance.create.schema.try(:to_json) rescue nil` 

is more idiomatic and then you don't need the other 4 lines

File: lib/services/service_brokers/service_manager.rb [hunk 2]
```diff
@@ -58,16 +58,34 @@ def update_or_create_plans(catalog)
...
+        rescue
+        end
```
Comment (ericpromislow) [inline lib/services/service_brokers/service_manager.rb]: There's a potential for exceptions to be lost here -- in the old code they would have been propagated to the top. Here the exceptions should be logged.

Also, in this block there's no need for the `.try` component.

File: lib/services/service_brokers/v2/catalog_plan.rb [hunk 3]
```diff
@@ -37,10 +46,13 @@ def validate!
...
       validate_bool!(:free, free) if free
       validate_bool!(:bindable, bindable) if bindable
+      validate_hash!(:schemas, @schemas_data) if @schemas_data
     end
 
     def validate_schemas!
-      errors.add_nested(schemas, schemas.errors) unless schemas.valid?
+      if schemas
+        errors.add_nested(schemas, schemas.errors) unless schemas.valid?
+      end
```
Comment (ericpromislow) [inline lib/services/service_brokers/v2/catalog_plan.rb]: I would simplify this as

    if schemas && !schemas.valid?  
      errors.add_nested(schemas, schemas.errors)
    end

File: lib/services/service_brokers/v2/parameters_schema.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      @errors = VCAP::Services::ValidationErrors.new
+      @parameters = parameters
+      @parameters_data = parameters['parameters']
+      @path = path + ['parameters']
+
+      build_schema
+    end
+
+    def build_schema
+      return unless @parameters_data
```
Comment (ericpromislow) [inline lib/services/service_brokers/v2/parameters_schema.rb]: No need for the first test because `nil.is_a?(Hash)` => `false`

File: lib/services/service_brokers/v2/service_instance_schema.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      @update_data = instance['update']
+
+      @path = ['service_instance']
+
+      build_create
+      build_update
+    end
+
+    def build_create
+      return unless @create_data
```
Comment (ericpromislow) [inline lib/services/service_brokers/v2/service_instance_schema.rb]: `return unless @create_data` not needed.

**Comment by ericpromislow** [comment]
Bunch of minor points addressed inline.

**Comment by lurraca** [comment]
We made some changes to the content of this PR. Please let us know what you think.

Luis & @AlbertoImpl

**Comment by matt-royal** [comment]
Hey @lurraca. The changes look good to us. Please resolve the conflicts and then we'll merge this.

Matt & @anniesing

**Comment by utako** [comment]
Hi @matt-royal and @anniesing!

We've resolved all of the conflicts and it should be good to go.

Thanks,
@ablease and @utako

File: lib/services/service_brokers/service_manager.rb [hunk 6]
```diff
@@ -58,16 +59,37 @@ def update_or_create_plans(catalog)
           plan.public = false
         end
 
+        create_instance = nil
+        begin
+          create_instance = catalog_plan.schemas.service_instance.create.schema.to_json
+        rescue => e
+          @logger.error("Error parsing service_instance create schema: #{e.message}")
+        end
+
```
Comment (ljfranklin) [inline lib/services/service_brokers/service_manager.rb]: I'm concerned about these rescues as the intent is unclear. Do you expect one of the fields to be nil and rescue a nil pointer exception? Do you expect a JSON marshal error due to bad input? Is there error actually cosmetic or was there a misunderstanding that `@logger.error` would re-raise?

Could you rewrite this to be more explicit about which failure mode you expect?

**Comment by ljfranklin** [comment]
Hey @ablease & @utako, added one more ask inline about the `rescue => e` blocks. Sorry about the delay in getting this merged, we're working on ironing out our community process to be a little more responsive going forward. Thanks for your patience!

File: lib/services/service_brokers/service_manager.rb (hunk 6 above)
Comment (utako) [reply lib/services/service_brokers/service_manager.rb]: As far as I can tell, we're just protecting against the absence of create_instance, update_instance, and/or create_binding schema (e.g. if catalog_plan.schemas.service_instance is nil, etc.). We shouldn't actually expect any exceptions aside from one that can be avoided by a conditional and some `try`ing.

**Comment by tcdowney** [comment]
Hey @ablease and @utako, we're taking another look at this PR since it looks like you've added commits since Lyle's comment.  There's quite a few of them though, could you respond to his comment with a commit SHA if one of those addresses it?  That could also be helpful going forward to speed these up. 👍

File: lib/services/service_brokers/service_manager.rb (hunk 6 above)
Comment (tcdowney) [reply lib/services/service_brokers/service_manager.rb]: Looks like this was updated in this commit: 804050c91ee05d425b4637d85c8d451732df87c6

Comment (Samze) [reply lib/services/service_brokers/service_manager.rb]: We updated this and did some refactoring in 1046fd554920502ef2286d00491358f8e3bfcf14 

We no longer use rescue as a control flow instead we use try. I hope this makes the code a bit clearer.

**Comment by Samze** [comment]
Thanks @tcdowney. We have added more commits and replied to @ljfranklin
